libio/serial_com.h

```diff
@@ -47,7 +47,7 @@ class SerialCom
 	void setVerbose(bool verb) { verbose = verb; }
 	void connect(const std::string &sDevice);
 	void disconnect();
-	void flush();
+	void flush(const std::chrono::milliseconds &timeout = std::chrono::seconds(1));
```

Comment:
@guihomework, I guess this is what you asked for.

Reply:
Yes but there are issues with it due to the `read timeout` (set by the function `setTimeout`) set in the ROS node to 1000 ms so the flush throws the `flush timeout` message even if there is no buffer to flush.

```
connecting to /dev/ttyACM0
connected
creating serial protocol with device_filename (...)
initializing serial protocol
Device didn't stop streaming. Try to disconnect and reconnect.
sp: init failed:sp: failed to stop streaming
initialization failed
disconnecting
disconnected
```

I tested with summing the `read timeout` to your `flush timeout` to have a correct deadline and this then worked for both clean start of a device (empty buffer already) and crashed start (=still streaming).

```
(...) device still streaming and stop issued, currently flushing
sc: trying to read 256 bytes
sc: pselect result 1 
sc: reading 256 bytes
sc: read result 16 
sc: pselect result 1 
sc: reading 240 bytes
sc: read result 240 
sc: read total of 256 bytes
sc: trying to read 256 bytes
sc: pselect result 1 
sc: reading 256 bytes
sc: read result 256 
sc: read total of 256 bytes
sc: trying to read 256 bytes
sc: pselect result 1 
sc: reading 256 bytes
sc: read result 24 
sc: pselect result 0 
sc: flushed 24 chars in 1002 ms    
```
the 1002 ms is 1000 ms of `read_timeout` + 2 ms to flush 7 * 256 bytes of data after the stop was issued

with my wrong glove not using SP I get
``` 
(...)
sc: read total of 256 bytes
sc: flushed 256 chars in 2000 ms
Device didn't stop streaming. Try to disconnect and reconnect.
sp: init failed:sp: failed to stop streaming
initialization failed
disconnecting
```
as expected.

But I think 1 sec `read_timeout` + 1 second `flush_timeout` is too long, as the flushing should not be more than a few milliseconds to add on top of the `read_timeout` until the buffer is empty (2 ms in case of the crashed myrmex_v2)

my solution is the quickest not changing the other `read_timeout` to std::chrono to not risk adding more changes to the system now. 
please ignore my wrong verbose message, it is just there to see how long flushing takes, the len is obviously not the total flushed which is not so interesting anyway.

```
@@ -182,12 +182,14 @@ size_t SerialCom::write(const uint8_t *buf, size_t len)
        return res;
 }

-void SerialCom::flush(const std::chrono::milliseconds &timeout)
+void SerialCom::flush(const std::chrono::milliseconds &flush_timeout)
 {
        unsigned char buf[256];
        size_t len = sizeof(buf);
        auto now = std::chrono::steady_clock::now();
-       auto deadline = now + timeout;
+       auto read_timeout = std::chrono::seconds{timeout.tv_sec} + std::chrono::nanoseconds{timeout.tv_nsec};
+
+       auto deadline = now + read_timeout + flush_timeout;
        try
        {
                // read remaining buffer
@@ -196,11 +198,16 @@ void SerialCom::flush(const std::chrono::milliseconds &timeout)
                        len = read(buf, sizeof(buf));
                        now = std::chrono::steady_clock::now();
                }
+               if (verbose)
+               {
+                       auto milliseconds = std::chrono::duration_cast<std::chrono::milliseconds>(now - (deadline - (read_timeout + flush_timeout)));
+                       printf("sc: flushed %lu chars in %ld ms\n" , len, milliseconds.count());
+               }
        }
        catch (const std::exception &e) {
                std::cerr << e.what() << std::endl;
        }
-       if (now >= deadline)
+       if (len > 0 && now >= deadline)
                throw std::runtime_error("Device didn't stop streaming. Try to disconnect and reconnect.");
 }
```

Reply:
You ask me something I don't remember and I should have written in the comment to justify but I did not. Usually I have good reasons. The git blame shows the node derives from a test app on a teensy with an IMU. Maybe the IMU initialization took time after the serial port opened, and the read was then taking time. In teensy I remember some firmware did not initialize other peripherals before the serial port was open (showing the device should start doing things). So this timeout might be related to this IMU OR something else.  Changing it would require to verify we do not break any device. 
I know code should not contain meaningless bits, or strange ones, but changing it requires somebody to test, and there is nobody to do that job anymore. I don't have time to keep testing any change done to the driver. Took me already too much time today to verify why the flush was not working, although it looked well on code reviewing. 
One should maybe think of a simulation of a serial device with various fail cases and see how the driver reacts to see if changes break the test (including timeouts etc spurious disconnections).

Reply:
> One should maybe think of a simulation of a serial device with various fail cases

Simulating (as well as testing) a device will always be limited to the failure cases considered at design time (or at test time). Sure, that might help to cover typical issues, but both are not bullet-proof.